File: app/routes/api_v2/schemas/exam_import.py
"""Deneme PDF içe aktarma — API v2 şemaları (önizleme taslağı + onay)."""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ImportDraftRow(BaseModel):
    """Önizleme tablosunun bir soru satırı — her alan düzenlenebilir."""
    exam_part: str | None = None          # birleşik belgede oturum: tyt|ayt
    subject_raw: str | None = None
    subject_id: int | None = None
    subject_name: str | None = None
    # okul-müfredat sınavında SUNUM grubu (sınıf dersinin adı — "TDE 21 +
    # TYT Türkçe 9" bölünmesi yerine tek "Türk Dili ve Edebiyatı"); None = yok
    display_subject: str | None = None
    question_no: int | None = None
    topic_raw: str | None = None
    topic_id: int | None = None
    topic_name: str | None = None
    topic_source: str | None = None       # alias|auto|ai|none
    correct_answer: str | None = None
    student_answer: str | None = None
    result: str | None = None             # dogru|yanlis|bos (None → şüpheli)
    is_suspect: bool = False


class ImportDraftSubject(BaseModel):
    name: str
    part: str | None = None               # birleşik belgede oturum: tyt|ayt
    questions: int
    correct: int
    wrong: int
    blank: int
    net: float
    doc_net: float | None = None          # belge kendi neti (çapraz kıyas)


class ImportPart(BaseModel):
    """Birleşik belgede (TG kitapçığı) bir sınav oturumu — koç birini seçip
    kaydeder; ikisini isterse iki ayrı içe aktarma yapar (kredi 1 kez düşer,
    çünkü analiz tek)."""
    part: str | None = None               # tyt|ayt (tek sınavlı belgede None)
    section: str
    section_label: str
    question_count: int


class ImportCheck(BaseModel):
    code: str
    label: str
    ok: bool
    detail: str | None = None


class ImportMatchStats(BaseModel):
    alias: int
    auto: int
    ai: int
    none: int


class SectionChoice(BaseModel):
    value: str
    label: str


class TopicChoice(BaseModel):
    """Önizlemede konu düzeltme seçicisinin adayları (evrenin resmi konuları)."""
    id: int
    name: str
    subject_name: str


class ExamImportDraft(BaseModel):
    """analyze çıktısı — önizleme ekranının tamamı."""
    title: str | None
    exam_date: str | None
    grade_hint: int | None
    universe: str                          # tyt|ayt|lgs|okul
    section: str
    section_label: str
    scope: str                             # full|brans
    confidence: str                        # high|medium|low
    parts: list[ImportPart]                # >1 ise birleşik belge (oturum seçici)
    subjects: list[ImportDraftSubject]
    rows: list[ImportDraftRow]
    checks: list[ImportCheck]
    suspect_count: int
    match_stats: ImportMatchStats
    duplicate_exam_id: int | None
    score_info: dict[str, Any] | None
    topic_choices: list[TopicChoice]       # konu düzeltme seçicisi adayları
    section_choices: list[SectionChoice]   # tür seçici (düşük güvende öne çıkar)
    credits_charged: int


class ConfirmRow(BaseModel):
    """Onay gövdesindeki satır — önizlemede düzeltilmiş hali."""
    subject_raw: str | None = None
    question_no: int | None = None
    topic_raw: str | None = None
    topic_id: int | None = None
    correct_answer: str | None = Field(default=None, max_length=8)
    student_answer: str | None = Field(default=None, max_length=8)
    result: str                            # dogru|yanlis|bos
    is_suspect: bool = False
    manually_edited: bool = False


class ExamImportConfirmBody(BaseModel):
    title: str
    exam_date: str                         # ISO YYYY-MM-DD
    section: str
    scope: str | None = None
    grade_hint: int | None = None
    note: str | None = Field(default=None, max_length=500)
    force: bool = False                    # mükerrer uyarısına rağmen kaydet
    score_info: dict[str, Any] | None = None
    rows: list[ConfirmRow]


class AnalysisSectionOption(BaseModel):
    value: str
    label: str
    count: int                             # soru-satırlı deneme sayısı


class AnalysisExamMeta(BaseModel):
    id: int
    title: str
    exam_date: str
    net: float


class AnalysisCell(BaseModel):
    """Isı haritası hücresi — bir konunun BİR denemedeki performansı."""
    exam_id: int
    total: int
    correct: int
    wrong: int
    blank: int
    accuracy: float                        # 0..1


class AnalysisTopicRow(BaseModel):
    topic_id: int
    topic_name: str
    subject_name: str
    total: int
    correct: int
    wrong: int
    blank: int
    accuracy: float
    exams_seen: int
    cells: list[AnalysisCell]


class AnalysisOpportunity(BaseModel):
    """Net fırsatı: bu konu kapanırsa deneme başına kazanılabilecek net."""
    topic_id: int
    topic_name: str
    subject_name: str
    total: int
    wrong: int
    blank: int
    accuracy: float
    net_gain_per_exam: float


class AnalysisTrendTopic(BaseModel):
    """Unutulan/gelişen konu — ilk yarı ↔ son yarı doğruluk kıyası."""
    topic_id: int
    topic_name: str
    subject_name: str
    first_accuracy: float
    last_accuracy: float


class ExamTopicAnalysisResponse(BaseModel):
    section: str | None
    section_label: str | None
    section_options: list[AnalysisSectionOption]
    exams: list[AnalysisExamMeta]
    topics: list[AnalysisTopicRow]         # ısı haritası satırları (soru sayısı DESC)
    opportunities: list[AnalysisOpportunity]
    forgotten: list[AnalysisTrendTopic]
    improved: list[AnalysisTrendTopic]
    unmatched_questions: int
    analyzed_question_count: int


class WrongBridgeResult(BaseModel):
    """Deneme → Yanlış Soru Arşivi köprüsü sonucu (Faz 3, idempotent)."""
    created: int
    skipped_existing: int                  # daha önce aktarılmış (mükerrer değil)
    skipped_no_topic: int                  # konusuz satır (önce 'Satırları düzelt')
    total_wrong: int


class ExamImportConfirmResult(BaseModel):
    exam_id: int
    title: str
    exam_date: str
    section: str
    section_label: str
    net: float
    total_correct: int
    total_wrong: int
    total_blank: int
    question_count: int
    matched_topic_count: int               # konuya bağlanan soru (birikime girer)
    wrong_topic_ids: list[int]             # Faz 3 YSA köprüsü için hazır veri

File: web/lib/types/exam-import.ts
/**
 * Deneme PDF içe aktarma — API v2 tipleri (schemas/exam_import.py aynası).
 */

export type ImportTopicSource = "alias" | "auto" | "ai" | "none" | "kayitli";
export type ImportResultValue = "dogru" | "yanlis" | "bos";

export interface ImportDraftRow {
  exam_part: string | null;
  subject_raw: string | null;
  subject_id: number | null;
  subject_name: string | null;
  /** Okul-müfredat sınavında sunum grubu (sınıf dersinin adı); yoksa null */
  display_subject?: string | null;
  question_no: number | null;
  topic_raw: string | null;
  topic_id: number | null;
  topic_name: string | null;
  topic_source: ImportTopicSource | null;
  correct_answer: string | null;
  student_answer: string | null;
  result: ImportResultValue | null;
  is_suspect: boolean;
}

export interface ImportDraftSubject {
  name: string;
  part: string | null;
  questions: number;
  correct: number;
  wrong: number;
  blank: number;
  net: number;
  doc_net: number | null;
}

export interface ImportPart {
  part: string | null;
  section: string;
  section_label: string;
  question_count: number;
}

export interface ImportCheck {
  code: string;
  label: string;
  ok: boolean;
  detail: string | null;
}

export interface ImportMatchStats {
  alias: number;
  auto: number;
  ai: number;
  none: number;
}

export interface SectionChoice {
  value: string;
  label: string;
}

export interface TopicChoice {
  id: number;
  name: string;
  subject_name: string;
}

export interface ExamImportDraft {
  title: string | null;
  exam_date: string | null;
  grade_hint: number | null;
  universe: "tyt" | "ayt" | "lgs" | "okul";
  section: string;
  section_label: string;
  scope: "full" | "brans";
  confidence: "high" | "medium" | "low";
  parts: ImportPart[];
  subjects: ImportDraftSubject[];
  rows: ImportDraftRow[];
  checks: ImportCheck[];
  suspect_count: number;
  match_stats: ImportMatchStats;
  duplicate_exam_id: number | null;
  score_info: Record<string, unknown> | null;
  topic_choices: TopicChoice[];
  section_choices: SectionChoice[];
  credits_charged: number;
}

export interface ConfirmRow {
  subject_raw: string | null;
  question_no: number | null;
  topic_raw: string | null;
  topic_id: number | null;
  correct_answer: string | null;
  student_answer: string | null;
  result: ImportResultValue;
  is_suspect: boolean;
  manually_edited: boolean;
}

export interface ExamImportConfirmBody {
  title: string;
  exam_date: string;
  section: string;
  scope?: string | null;
  grade_hint?: number | null;
  note?: string | null;
  force?: boolean;
  score_info?: Record<string, unknown> | null;
  rows: ConfirmRow[];
}

export interface ExamImportConfirmResult {
  exam_id: number;
  title: string;
  exam_date: string;
  section: string;
  section_label: string;
  net: number;
  total_correct: number;
  total_wrong: number;
  total_blank: number;
  question_count: number;
  matched_topic_count: number;
  wrong_topic_ids: number[];
}

// ============================================================================
// Konu × deneme analizi (Faz 2)
// ============================================================================

export interface AnalysisSectionOption {
  value: string;
  label: string;
  count: number;
}

export interface AnalysisExamMeta {
  id: number;
  title: string;
  exam_date: string;
  net: number;
}

export interface AnalysisCell {
  exam_id: number;
  total: number;
  correct: number;
  wrong: number;
  blank: number;
  accuracy: number;
}

export interface AnalysisTopicRow {
  topic_id: number;
  topic_name: string;
  subject_name: string;
  total: number;
  correct: number;
  wrong: number;
  blank: number;
  accuracy: number;
  exams_seen: number;
  cells: AnalysisCell[];
}

export interface AnalysisOpportunity {
  topic_id: number;
  topic_name: string;
  subject_name: string;
  total: number;
  wrong: number;
  blank: number;
  accuracy: number;
  net_gain_per_exam: number;
}

export interface AnalysisTrendTopic {
  topic_id: number;
  topic_name: string;
  subject_name: string;
  first_accuracy: number;
  last_accuracy: number;
}

export interface ExamTopicAnalysisResponse {
  section: string | null;
  section_label: string | null;
  section_options: AnalysisSectionOption[];
  exams: AnalysisExamMeta[];
  topics: AnalysisTopicRow[];
  opportunities: AnalysisOpportunity[];
  forgotten: AnalysisTrendTopic[];
  improved: AnalysisTrendTopic[];
  unmatched_questions: number;
  analyzed_question_count: number;
}

/** Seçici köprü: denemenin bir yanlış sorusu (arşiv durumu işaretli). */
export interface ExamWrongRow {
  question_id: number;
  question_no: number | null;
  subject: string | null;
  topic_id: number | null;
  topic_name: string | null;
  topic_label_raw: string | null;
  correct_answer: string | null;
  student_answer: string | null;
  archived: boolean;
}

export interface ExamWrongRowsResponse {
  exam_id: number;
  title: string;
  rows: ExamWrongRow[];
  error_types: { value: string; label: string }[];
}

export interface WrongBridgeItem {
  question_id: number;
  error_type?: string | null;
}

/** İçe aktarmada kullanıcı beyanı (BEYAN ESAS — tespit bekçiye döner). */
export interface ImportDeclaration {
  declaredSection?: string | null;
  declaredGrade?: number | null;
}

/** Deneme → Yanlış Soru Arşivi köprüsü sonucu (Faz 3, idempotent). */
export interface WrongBridgeResult {
  created: number;
  skipped_existing: number;
  skipped_no_topic: number;
  total_wrong: number;
}
